### PR TITLE
Datashader support for points and shapes

### PR DESCRIPTION
**Pull request by LucaMarconato**
Work from @Sonja-Stockhaus and @ckmah.

### CODE REVIEW

**Comment by LucaMarconato** [comment]
@Sonja-Stockhaus recently we introduced datashader support in `spatialdata` for points and shapes and you may find it useful to reuse some components. 

In particular, you may find it useful to reuse the code in https://github.com/scverse/spatialdata/blob/b8fbc5c4ce7643455768a94665496ea4987294dd/src/spatialdata/_core/operations/rasterize.py#L604. 

Technically, I would consider isolating the code used to compute `cnv = ds.Canvas(...)` into a private function of `rasterize.py` in `spatialdata`, and then I would use that private function from `spatialdata-plot` (using `rasterize_shapes_points()` directly may not be useful for `spatialdata-plot` since you probably want more control on the shading operation, so reusing some intermediate code by placing it on private functions may be the way to go).

Please let me know what you think about this. Also, if you prefer to merge this PR as it is and then eventually reuse the components from `rasterize.py` in a follow up refactoring, that would also be perfectly fine.

**Comment by Sonja-Stockhaus** [comment]
> @Sonja-Stockhaus recently we introduced datashader support in `spatialdata` for points and shapes and you may find it useful to reuse some components.
> 
> In particular, you may find it useful to reuse the code in [scverse/spatialdata@`b8fbc5c`/src/spatialdata/_core/operations/rasterize.py#L604](https://github.com/scverse/spatialdata/blob/b8fbc5c4ce7643455768a94665496ea4987294dd/src/spatialdata/_core/operations/rasterize.py#L604).
> 
> Technically, I would consider isolating the code used to compute `cnv = ds.Canvas(...)` into a private function of `rasterize.py` in `spatialdata`, and then I would use that private function from `spatialdata-plot` (using `rasterize_shapes_points()` directly may not be useful for `spatialdata-plot` since you probably want more control on the shading operation, so reusing some intermediate code by placing it on private functions may be the way to go).
> 
> Please let me know what you think about this. Also, if you prefer to merge this PR as it is and then eventually reuse the components from `rasterize.py` in a follow up refactoring, that would also be perfectly fine.

Thanks for the heads up! I think fo now it would be best to finish and merge this PR and then do the refactoring later, since it's already so big.
Also regarding that @LucaMarconato can you tell me when the next pip release of `spatialdata` will be? Since this code is currently using some of the newer changes and that's why the tests keep failing right now

**Comment by LucaMarconato** [comment]
Your message comes at the right time 😊 I just made a new release 10 minutes ago!

I agree, let's rather merge this and refactor later.

**Comment by timtreis** [comment]
@Sonja-Stockhaus is this ready for review?

File: src/spatialdata_plot/pl/render.py [hunk 1]
```diff
@@ -171,6 +146,130 @@ def _render_shapes(
...
+    # Determine which method to use for rendering
+    method = render_params.method
+    if method is None:
+        method = "datashader" if len(shapes) > 10000 else "matplotlib"
+    elif method not in ["matplotlib", "datashader"]:
+        raise ValueError("Method must be either 'matplotlib' or 'datashader'.")
+    logger.info(f"Using {method}")
+
+    if method == "datashader":
+        # TODO: Where to put this
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:165]: wdym?

File: src/spatialdata_plot/pl/render.py [hunk 2]
```diff
@@ -171,6 +146,130 @@ def _render_shapes(
...
+        is_point = _geometry.type == "Point"
+
+        # Handle circles encoded as points with radius
+        if is_point.any():
+            scale = shapes[is_point]["radius"] * render_params.scale
+            sdata_filt.shapes[element].loc[is_point, "geometry"] = _geometry[is_point].buffer(scale)
+
+        # in case we are coloring by a column in table
+        if col_for_color is not None and col_for_color not in sdata_filt.shapes[element].columns:
+            # numerical
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:189]: remove or be more explicit

File: src/spatialdata_plot/pl/render.py [hunk 3]
```diff
@@ -171,6 +146,130 @@ def _render_shapes(
...
+        # Render shapes with datashader
+        color_by_categorical = col_for_color is not None and color_source_vector is not None
+        aggregate_with_sum = None
+        if col_for_color is not None and (render_params.groups is None or len(render_params.groups) > 1):
+            if color_by_categorical:
+                agg = cvs.polygons(
+                    sdata_filt.shapes[element], geometry="geometry", agg=ds.by(col_for_color, ds.count())
+                )
+            else:
+                # numerical
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:202]: remove or be more explicit

File: src/spatialdata_plot/pl/render.py [hunk 4]
```diff
@@ -171,6 +146,130 @@ def _render_shapes(
...
+                color_key=color_key,
+                min_alpha=np.min([150, render_params.fill_alpha * 255]),
+            )
+            if color_by_categorical or col_for_color is None
+            else ds.tf.shade(
+                agg,
+                cmap=render_params.cmap_params.cmap,
+            )
+        )
+        # Render image
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:229]: remove

File: src/spatialdata_plot/pl/render.py [hunk 5]
```diff
@@ -236,6 +335,9 @@ def _render_points(
 
     if groups is not None and col_for_color is not None:
         points = points[points[col_for_color].isin(groups)]
+        # in case no rows are left:
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:338]: remove

File: src/spatialdata_plot/pl/render.py [hunk 6]
```diff
@@ -291,19 +398,109 @@ def _render_points(
...
+            if ax.yaxis_inverted():  # case for e.g. images
+                y_ext = [min(y_ext[0], previous_ylim[1]), max(y_ext[1], previous_ylim[0])]
+            else:  # case for e.g. labels
+                y_ext = [min(y_ext[0], previous_ylim[0]), max(y_ext[1], previous_ylim[1])]
+        # round because we need integers
+        plot_width = int(np.round(x_ext[1] - x_ext[0]))
+        plot_height = int(np.round(y_ext[1] - y_ext[0]))
+
+        # use datashader for the visualization of points
+        # TODO: what about trans/norm at this point?
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:438]: wdym?

File: src/spatialdata_plot/pl/render.py [hunk 7]
```diff
@@ -291,19 +398,109 @@ def _render_points(
...
+        cvs = ds.Canvas(plot_width=plot_width, plot_height=plot_height, x_range=x_ext, y_range=y_ext)
+
+        color_by_categorical = col_for_color is not None and points[col_for_color].values.dtype == object
+        aggregate_with_sum = None
+        if col_for_color is not None and (render_params.groups is None or len(render_params.groups) > 1):
+            if color_by_categorical:
+                agg = cvs.points(sdata_filt.points[element], "x", "y", agg=ds.by(col_for_color, ds.count()))
+            else:
+                # numerical
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:447]: remove or be more explicit

File: src/spatialdata_plot/pl/render.py [hunk 8]
```diff
@@ -291,19 +398,109 @@ def _render_points(
...
+            else None
+        )
+        if color_by_categorical or col_for_color is None:
+            ds_result = ds.tf.shade(
+                ds.tf.spread(agg, px=px),
+                rescale_discrete_levels=True,
+                cmap=color_vector[0][:-2],
+                color_key=color_key,
+                min_alpha=np.min([150, render_params.alpha * 255]),
+            )  # TODO: choose other value than 150 for min_alpha (here and below)?
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:467]: Why a different one? Is 150 bad?

File: src/spatialdata_plot/pl/render.py [hunk 9]
```diff
@@ -437,10 +633,13 @@ def _render_images(
                     clip=True,
                 )
 
-            if not isinstance(render_params.cmap_params, list) and render_params.cmap_params.norm:
-                layers[c] = render_params.cmap_params.norm(layers[c])
-            elif isinstance(render_params.cmap_params, list) and render_params.cmap_params[ch_index].norm:
-                layers[c] = render_params.cmap_params[ch_index].norm(layers[c])
+            # TODO: can't be list anymore???
```
Comment (timtreis) [inline src/spatialdata_plot/pl/render.py:636]: wdym?

**Review by timtreis — APPROVED** [review]
Already looks fairly good! Can we also add examples to your speed improvement notebook to show these?

File: src/spatialdata_plot/pl/render.py (hunk 8 above)
Comment (Sonja-Stockhaus) [reply src/spatialdata_plot/pl/render.py:467]: Not necessarily, but it's an arbitrary value that I chose to make everything better visible...

**Pull request merged**